rosidl python generator makes invalid assumptions about message generation target names

## Bug report

**Required Info:**

- Operating System:
  - N/A
- Installation type:
  - master
- Version or commit hash:
  - latest
- DDS implementation:
  - N/A
- Client library (if applicable):
  - rclpy

#### Steps to reproduce issue

Here's a simple package that causes the issue:

https://gist.github.com/wjwwood/72e16b014ef2f7bab56fea96af3c974e

#### Expected behavior

It configures and compiles.

#### Actual behavior

I get this error when configuring:

```
% cmake ..
-- Found ament_cmake: 0.5.1 (/Users/william/ros2_ws/install/ament_cmake/share/ament_cmake/cmake)
-- Found PythonInterp: /Users/william/.pyenv/shims/python3 (found suitable version "3.7", minimum required is "3")
-- Using PYTHON_EXECUTABLE: /Users/william/.pyenv/shims/python3
-- Found rosidl_default_generators: 0.5.0 (/Users/william/ros2_ws/install/rosidl_default_generators/share/rosidl_default_generators/cmake)
-- Using all available rosidl_typesupport_c: rosidl_typesupport_introspection_c;rosidl_typesupport_fastrtps_c
-- Using all available rosidl_typesupport_cpp: rosidl_typesupport_introspection_cpp;rosidl_typesupport_fastrtps_cpp
-- Found rmw_implementation_cmake: 0.5.0 (/Users/william/ros2_ws/install/rmw_implementation_cmake/share/rmw_implementation_cmake/cmake)
-- Found PythonInterp: /Users/william/.pyenv/shims/python3 (found suitable version "3.7", minimum required is "3.5")
-- Found python_cmake_module: 0.5.2 (/Users/william/ros2_ws/install/python_cmake_module/share/python_cmake_module/cmake)
-- Using PythonExtra_INCLUDE_DIRS: /Users/william/.pyenv/versions/3.7.0/Python.framework/Versions/3.7/include/python3.7m;/Users/william/.pyenv/versions/3.7.0/include/python3.7m
-- Using PythonExtra_LIBRARIES: /Users/william/.pyenv/versions/3.7.0/Python.framework/Versions/3.7/lib/python3.7/config-3.7m-darwin/libpython3.7m.dylib
CMake Error at /Users/william/ros2_ws/install/rosidl_cmake/share/rosidl_cmake/cmake/rosidl_target_interfaces.cmake:44 (message):
  rosidl_target_interfaces() the second argument 'test_pkg' concatenated with
  the third argument 'rosidl_typesupport_c' using double underscores must be
  a valid target name
Call Stack (most recent call first):
  /Users/william/ros2_ws/install/rosidl_generator_py/share/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake:241 (rosidl_target_interfaces)
  /Users/william/ros2_ws/install/ament_cmake_core/share/ament_cmake_core/cmake/core/ament_execute_extensions.cmake:38 (include)
  /Users/william/ros2_ws/install/rosidl_cmake/share/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake:169 (ament_execute_extensions)
  CMakeLists.txt:17 (rosidl_generate_interfaces)


-- Configuring incomplete, errors occurred!
See also "/tmp/dejan/build/CMakeFiles/CMakeOutput.log".
```

#### Additional information

I believe that this stems from the Python rosidl generator assuming that `PROJECT_NAME` is always the first part of the rosidl generation target name, but in this example it is `${PROJECT_NAME}_ext_to_avoid_target_collision` instead (normally it is just `${PROJECT_NAME}`).

This is sometimes necessary when you want to have another target be just `${PROJECT_NAME}`, and so renaming the `rosidl` target is necessary.

--- a/rosidl_generator_py/CMakeLists.txt
+++ b/rosidl_generator_py/CMakeLists.txt
@@ -62,7 +62,10 @@ if(BUILD_TESTING)
   if(_typesupport_impls STREQUAL "")
     message(WARNING "No valid typesupport found for Python generator, skipping tests.")
   else()
-    rosidl_generate_interfaces(${PROJECT_NAME} ${message_files}
+    # Use a custom suffix for the target name here to detect regressions of this:
+    #   https://github.com/ros2/rosidl_python/issues/11
+    # The alternative style (just using ${PROJECT_NAME}) is used by most interface packages.
+    rosidl_generate_interfaces(${PROJECT_NAME}_custom ${message_files}
       ADD_LINTER_TESTS
       SKIP_INSTALL
     )

--- a/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
+++ b/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
@@ -228,7 +228,7 @@ foreach(_typesupport_impl ${_typesupport_impls})
   target_link_libraries(
     ${_target_name}
     ${PythonExtra_LIBRARIES}
-    ${PROJECT_NAME}__${_typesupport_impl}
+    ${rosidl_generate_interfaces_TARGET}__${_typesupport_impl}
   )
 
   target_include_directories(${_target_name}
@@ -239,7 +239,7 @@ foreach(_typesupport_impl ${_typesupport_impls})
   )
 
   rosidl_target_interfaces(${_target_name}
-    ${PROJECT_NAME} rosidl_typesupport_c)
+    ${rosidl_generate_interfaces_TARGET} rosidl_typesupport_c)
 
   ament_target_dependencies(${_target_name}
     "rosidl_generator_c"
@@ -258,7 +258,7 @@ foreach(_typesupport_impl ${_typesupport_impls})
   ament_target_dependencies(${_target_name}
     "rosidl_generator_c"
     "rosidl_generator_py"
-    "${PROJECT_NAME}__rosidl_generator_c"
+    "${rosidl_generate_interfaces_TARGET}__rosidl_generator_c"
   )
   set(PYTHON_EXECUTABLE ${_PYTHON_EXECUTABLE})
 

